File: CommonDependencies/dependencies/py/promql_utilities/promql_utilities/query_logics/logics.py
from typing import Tuple

from promql_utilities.query_logics.enums import QueryTreatmentType, Statistic

# def map_statistic_to_precompute_operators(
#     statistic: str, treatment_type: QueryTreatmentType
# ) -> List[Tuple[str, str]]:
#     # if statistic in ["quantile", "stddev", "stdvar"]:
#     if statistic == "quantile":
#         if treatment_type == QueryTreatmentType.EXACT:
#             raise ValueError(f"Statistic {statistic} cannot be computed exactly")
#         else:
#             return [("KLL", "")]
#         # else:
#         #     return [("UnivMon", "")]
#     elif statistic in ["min", "max"]:
#         if treatment_type == QueryTreatmentType.APPROXIMATE:
#             return [("KLL", "")]
#         else:
#             return [("MinMax", statistic)]
#     elif statistic in ["sum", "count"]:
#         if treatment_type == QueryTreatmentType.APPROXIMATE:
#             return [("CountMinSketch", statistic)]
#         else:
#             return [("Sum", statistic)]
#     elif statistic == "avg":
#         if treatment_type == QueryTreatmentType.APPROXIMATE:
#             return [("CountMinSketch", "sum"), ("CountMinSketch", "count")]
#         else:
#             return [("Sum", "sum"), ("Sum", "count")]
#     elif statistic in ["rate", "increase"]:
#         return [("Increase", "")]
#     else:
#         raise NotImplementedError(f"Statistic {statistic} not supported")


def map_statistic_to_precompute_operator(
    statistic: Statistic, treatment_type: QueryTreatmentType
) -> Tuple[str, str]:
    # if statistic in ["quantile", "stddev", "stdvar"]:
    if statistic == Statistic.QUANTILE:
        if treatment_type == QueryTreatmentType.EXACT:
            raise ValueError(f"Statistic {statistic} cannot be computed exactly")
        else:
            return ("DatasketchesKLL", "")
            # return ("HydraKLL", "")
        # else:
        #     return [("UnivMon", "")]
    elif statistic == Statistic.TOPK:
        if treatment_type == QueryTreatmentType.EXACT:
            raise ValueError(f"Statistic {statistic} cannot be computed exactly")
        else:
            return ("CountMinSketchWithHeap", statistic.name.lower())
    elif statistic in [Statistic.MIN, Statistic.MAX]:
        if treatment_type == QueryTreatmentType.APPROXIMATE:
            return ("DatasketchesKLL", "")
            # return ("HydraKLL", "")
        else:
            # NOTE: Change to Multiple<>Accumulator
            # return ("MinMax", statistic.name.lower())
            return ("MultipleMinMax", statistic.name.lower())
    elif statistic in [Statistic.SUM, Statistic.COUNT]:
        if treatment_type == QueryTreatmentType.APPROXIMATE:
            return ("CountMinSketch", statistic.name.lower())
        else:
            # NOTE: Change to Multiple<>Accumulator
            # return ("Sum", statistic.name.lower())
            return ("MultipleSum", statistic.name.lower())
    # elif statistic == "avg":
    #     if treatment_type == QueryTreatmentType.APPROXIMATE:
    #         return [("CountMinSketch", "sum"), ("CountMinSketch", "count")]
    #     else:
    #         return [("Sum", "sum"), ("Sum", "count")]
    elif statistic in [Statistic.RATE, Statistic.INCREASE]:
        # NOTE: Change to Multiple<>Accumulator
        # return ("Increase", "")
        return ("MultipleIncrease", "")
    else:
        raise NotImplementedError(f"Statistic {statistic} not supported")


def does_precompute_operator_support_subpopulations(
    statistic: Statistic, precompute_operator: str
) -> bool:
    if precompute_operator in ["Increase", "MinMax", "Sum", "DatasketchesKLL"]:
        return False
    elif precompute_operator in [
        "MultipleIncrease",
        "MultipleMinMax",
        "MultipleSum",
        "HydraKLL",
    ]:
        # TODO: do we need to check for statistic here? If not, remove the check from CountMinSketch
        return True
    elif precompute_operator == "CountMinSketch":
        return statistic in [Statistic.SUM, Statistic.COUNT]
    elif (
        precompute_operator == "CountMinSketchWithHeap" and statistic == Statistic.TOPK
    ):
        # topk and bottomk do not support subpopulations!
        # other usages of CountMinSketchWithHeap will fall through.
        return False
    # elif precompute_operator == "UnivMon":
    #     return statistic in ["sum", "count", "avg"]
    else:
        raise NotImplementedError(
            f"Precompute operator {precompute_operator} not supported"
        )


def get_is_collapsable(temporal_aggregation: str, spatial_aggregation: str) -> bool:
    if spatial_aggregation == "sum":
        return temporal_aggregation in [
            "sum_over_time",
            "count_over_time",
            # "increase",
            # "rate",
        ]
    elif spatial_aggregation == "min":
        return temporal_aggregation == "min_over_time"
    elif spatial_aggregation == "max":
        return temporal_aggregation == "max_over_time"
    return False

File: CommonDependencies/dependencies/rs/promql_utilities/src/query_logics/logics.rs
use crate::query_logics::enums::{QueryTreatmentType, Statistic};
use tracing::debug;

/// Map statistic to precompute operator based on treatment type
/// This mirrors the Python implementation's logic
pub fn map_statistic_to_precompute_operator(
    statistic: Statistic,
    treatment_type: QueryTreatmentType,
) -> Result<(String, String), String> {
    debug!(
        "Mapping statistic {:?} with treatment type {:?} to precompute operator",
        statistic, treatment_type
    );
    match statistic {
        Statistic::Quantile => {
            if treatment_type == QueryTreatmentType::Exact {
                Err("Statistic Quantile cannot be computed exactly".to_string())
            } else {
                Ok(("DatasketchesKLL".to_string(), "".to_string()))
                //Ok(("HydraKLL".to_string(), "".to_string()))
            }
        }
        Statistic::Min | Statistic::Max => {
            if treatment_type == QueryTreatmentType::Approximate {
                Ok(("DatasketchesKLL".to_string(), "".to_string()))
                //Ok(("HydraKLL".to_string(), "".to_string()))
            } else {
                Ok((
                    "MultipleMinMax".to_string(),
                    statistic.to_string().to_lowercase(),
                ))
            }
        }
        Statistic::Sum | Statistic::Count => {
            if treatment_type == QueryTreatmentType::Approximate {
                Ok((
                    "CountMinSketch".to_string(),
                    statistic.to_string().to_lowercase(),
                ))
            } else {
                Ok((
                    "MultipleSum".to_string(),
                    statistic.to_string().to_lowercase(),
                ))
            }
        }
        Statistic::Rate | Statistic::Increase => {
            Ok(("MultipleIncrease".to_string(), "".to_string()))
        }
        _ => Err(format!("Statistic {statistic:?} not supported")),
    }
}

/// Check if a precompute operator supports subpopulations (multiple keys)
pub fn does_precompute_operator_support_subpopulations(
    statistic: Statistic,
    precompute_operator: &str,
) -> bool {
    debug!(
        "Checking if precompute operator '{}' supports subpopulations for statistic {:?}",
        precompute_operator, statistic
    );
    match precompute_operator {
        // Single-key operators
        "Increase" | "MinMax" | "Sum" | "DatasketchesKLL" => false,

        // Multi-key operators
        "MultipleIncrease" | "MultipleMinMax" | "MultipleSum" | "HydraKLL" => true,

        // CountMinSketch supports subpopulations only for certain statistics
        "CountMinSketch" => matches!(statistic, Statistic::Sum | Statistic::Count),

        // "CountMinSketchWithHeap" is only supported for Topk
        // Other usages of CountMinSketchWithHeap will fall through.
        "CountMinSketchWithHeap" if matches!(statistic, Statistic::Topk) => false,

        // Default: not supported
        _ => panic!("Unexpected precompute operator: {}", precompute_operator),
    }
}

/// Check if temporal and spatial aggregations are collapsible
/// Based on Python implementation in promql_utilities/query_logics/logics.py
pub fn get_is_collapsable(temporal_aggregation: &str, spatial_aggregation: &str) -> bool {
    debug!(
        "Checking if temporal aggregation '{}' and spatial aggregation '{}' are collapsable",
        temporal_aggregation, spatial_aggregation
    );
    match spatial_aggregation {
        "sum" => matches!(
            temporal_aggregation,
            "sum_over_time" | "count_over_time" // Note: "increase" and "rate" are commented out in Python
        ),
        "min" => temporal_aggregation == "min_over_time",
        "max" => temporal_aggregation == "max_over_time",
        _ => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_map_statistic_to_precompute_operator() {
        // Test exact sum
        let result =
            map_statistic_to_precompute_operator(Statistic::Sum, QueryTreatmentType::Exact)
                .unwrap();
        assert_eq!(result, ("MultipleSum".to_string(), "sum".to_string()));

        // Test approximate sum
        let result =
            map_statistic_to_precompute_operator(Statistic::Sum, QueryTreatmentType::Approximate)
                .unwrap();
        assert_eq!(result, ("CountMinSketch".to_string(), "sum".to_string()));

        // Test exact quantile (should fail)
        let result =
            map_statistic_to_precompute_operator(Statistic::Quantile, QueryTreatmentType::Exact);
        assert!(result.is_err());

        // Test approximate quantile
        let result = map_statistic_to_precompute_operator(
            Statistic::Quantile,
            QueryTreatmentType::Approximate,
        )
        .unwrap();
        assert_eq!(result, ("DatasketchesKLL".to_string(), "".to_string()));
        //assert_eq!(result, ("HydraKLL".to_string(), "".to_string()));
    }

    #[test]
    fn test_does_precompute_operator_support_subpopulations() {
        // Test MultipleSum supports subpopulations
        assert!(does_precompute_operator_support_subpopulations(
            Statistic::Sum,
            "MultipleSum"
        ));

        // Test DatasketchesKLL does not support subpopulations
        assert!(!does_precompute_operator_support_subpopulations(
            Statistic::Quantile,
            "DatasketchesKLL"
        ));

        // Test HydraKLL supports subpopulations
        assert!(does_precompute_operator_support_subpopulations(
            Statistic::Quantile,
            "HydraKLL"
        ));

        // Test CountMinSketch with valid statistic
        assert!(does_precompute_operator_support_subpopulations(
            Statistic::Sum,
            "CountMinSketch"
        ));
    }

    #[test]
    fn test_get_is_collapsable() {
        assert!(get_is_collapsable("sum_over_time", "sum"));
        assert!(get_is_collapsable("count_over_time", "sum"));
        assert!(get_is_collapsable("min_over_time", "min"));
        assert!(get_is_collapsable("max_over_time", "max"));
        assert!(!get_is_collapsable("min_over_time", "sum"));
        assert!(!get_is_collapsable("unknown", "sum"));
    }
}
